src/commonMain/kotlin/dev/snipme/highlights/Highlights.kt

```diff
@@ -60,22 +69,37 @@ class Highlights private constructor(
     }
 
     fun getHighlights(): List<CodeHighlight> {
-        val highlights = mutableListOf<CodeHighlight>()
         val structure = getCodeStructure()
-        with(structure) {
-            marks.forEach { highlights.add(ColorHighlight(it, theme.mark)) }
-            punctuations.forEach { highlights.add(ColorHighlight(it, theme.punctuation)) }
-            keywords.forEach { highlights.add(ColorHighlight(it, theme.keyword)) }
-            strings.forEach { highlights.add(ColorHighlight(it, theme.string)) }
-            literals.forEach { highlights.add(ColorHighlight(it, theme.literal)) }
-            annotations.forEach { highlights.add(ColorHighlight(it, theme.metadata)) }
-            comments.forEach { highlights.add(ColorHighlight(it, theme.comment)) }
-            multilineComments.forEach { highlights.add(ColorHighlight(it, theme.multilineComment)) }
-        }
-
-        emphasisLocations.forEach { highlights.add(BoldHighlight(it)) }
+        return constructHighlights(structure)
+    }
 
-        return highlights
+    fun getHighlightsAsync(listener: HighlightsResultListener) {
+        try {
+            val errorHandler = CoroutineExceptionHandler { _, exception ->
+                listener.onError(exception)
+            }
+
+            fun runJob() {
+                CoroutineScope(Dispatchers.Default + errorHandler).launch {
+                    val structure = getCodeStructure()
+                    val highlights = constructHighlights(structure)
```

Comment:
You should add `ensureActive` here and there in-between steps to stop the processing early if the coroutine was cancelled

Reply:
Nice catch! I think I was missing that during the development